ion-schema/src/isl/isl_type.rs

```diff
@@ -1,61 +1,130 @@
-use crate::isl::isl_constraint::IslConstraint;
+use crate::isl::isl_constraint::{IslConstraint, IslConstraintImpl};
 use crate::isl::isl_import::IslImportType;
+use crate::isl::IonSchemaLanguageVersion;
 use crate::result::{invalid_schema_error, invalid_schema_error_raw, IonSchemaResult};
 use ion_rs::value::owned::{text_token, Element};
 use ion_rs::value::{IonElement, IonStruct};
 
+/// Provides public facing APIs for constructing ISL types programmatically for ISL 1.0
+pub mod v_1_0 {
+    use crate::isl::isl_constraint::{IslConstraint, IslConstraintImpl};
+    use crate::isl::isl_type::{IslType, IslTypeImpl, IslTypeKind};
+
+    /// Creates a [IslType::Named] using the [IslConstraint] defined within it
+    pub fn named_type<A: Into<String>, B: Into<Vec<IslConstraint>>>(
+        name: A,
+        constraints: B,
+    ) -> IslType {
+        let constraints = constraints.into();
+        let isl_constraints: Vec<IslConstraintImpl> = constraints
+            .iter()
+            .map(|c| c.constraint.to_owned())
+            .collect();
+        IslType::new(
+            IslTypeKind::Named(IslTypeImpl::new(Some(name.into()), isl_constraints, None)),
+            constraints,
+        )
+    }
+
+    /// Creates a [IslType::Anonymous] using the [IslConstraint] defined within it
+    pub fn anonymous_type<A: Into<Vec<IslConstraint>>>(constraints: A) -> IslType {
+        let constraints = constraints.into();
+        let isl_constraints: Vec<IslConstraintImpl> = constraints
+            .iter()
+            .map(|c| c.constraint.to_owned())
+            .collect();
+        IslType::new(
+            IslTypeKind::Anonymous(IslTypeImpl::new(None, isl_constraints, None)),
+            constraints,
+        )
+    }
+}
+
+/// Provides public facing APIs for constructing ISL types programmatically for ISL 2.0
+pub mod v_2_0 {
+    use crate::isl::isl_constraint::IslConstraint;
+    use crate::isl::isl_type::{v_1_0, IslType};
+
+    /// Creates a [IslType::Named] using the [IslConstraint] defined within it
+    pub fn named_type<A: Into<String>, B: Into<Vec<IslConstraint>>>(
+        name: A,
+        constraints: B,
+    ) -> IslType {
+        v_1_0::named_type(name, constraints)
+    }
+
+    /// Creates a [IslType::Anonymous] using the [IslConstraint] defined within it
+    pub fn anonymous_type<A: Into<Vec<IslConstraint>>>(constraints: A) -> IslType {
+        v_1_0::anonymous_type(constraints)
+    }
+}
+
 /// Represents a type in an ISL schema.
 #[derive(Debug, Clone, PartialEq)]
-pub enum IslType {
-    Named(IslTypeImpl),
-    Anonymous(IslTypeImpl),
+pub struct IslType {
+    pub(crate) kind: IslTypeKind,
+    constraints: Vec<IslConstraint>,
 }
```

Comment:
I am confused about this change.

We now have `IslType` which contains a `Vec<IslConstraint>` and an `IslTypeKind`. The `IslTypeKind` is an enum where both variants contain `IslTypeImpl`, and then `IslTypeImpl` contains an optional name and a `Vec<IslConstraintImpl>`.

- What is the point of `IslTypeKind`? Why not just have `name: Option<String>` in either `IslType` or `IslTypeImpl` and get rid of the extra indirection?
- We already have the separate internal model. What exactly is gained by having both `IslType` and `IslTypeImpl`? How does it help hide version information?

Reply:
> What is the point of IslTypeKind? Why not just have name: Option<String> in either IslType or IslTypeImpl and get rid of the extra indirection?
I just tried to add wrapper (`IslType`) around what was already available previously. But It does seem that we can get rid of `IslTypeKind` in this case.

> We already have the separate internal model. What exactly is gained by having both IslType and IslTypeImpl? How does it help hide version information?
I am adding `IslTypeImpl` is using `IslConstraintImpl` to be store its constraints. Similar to `IslConstraint`, I have added `IslType` to provide methods based on versions through modules `v_1_0` and `v_2_0`.

Reply:
I have  removed `IslTypeKind` with https://github.com/amazon-ion/ion-schema-rust/pull/152/commits/24c20fb7ebeb6c68d48a9daba28b040aafe16b49